### PR TITLE
#635 Search in channel members

### PR DESCRIPTION
**Pull request by RkAiman**

### CODE REVIEW

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Parts/Header/MemberChannelRow.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+const { Text } = Typography;
+
+type Props = {
+  userId: string;
+};
+
+export default (props: Props) => {
+  const [avatar, name] = getChannelParts({
+    usersIds: [props.userId] || [],
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Parts/Header/MemberChannelRow.tsx:18]: const {avatar, name} = getUsersParts({options})

//Exec
const {avatar, name, user} = getUsersParts({options})

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx [hunk 2]
```diff
@@ -88,61 +103,29 @@ const ChannelMembersList: FC<Props> = props => {
...
+          limit < searchedUsers.length + 100 &&
+          setLimit(searchedUsers.length + 100)}
+      </div>
+      {channelMembers.length >= limit && (
         <Row align="middle" justify="center" gutter={[0, 16]}>
           <Link
             className="small-y-margin"
             style={{ color: 'var(--grey-dark)' }}
-            onClick={() => console.log('show list +5 members')}
+            onClick={() => setLimit(channelMembers.length + 100)}
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx:122]: Put the "100" into a const on top of functional component

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx [hunk 3]
```diff
@@ -1,50 +1,68 @@
...
   const collectionPath: string = `/channels/v1/companies/${company_id}/workspaces/${workspace_id}/channels/${id}/members/`;
   const channelMembersCollection = Collections.get(collectionPath, ChannelMemberResource);
+  const channelMembers = channelMembersCollection.useWatcher({}, { limit: limit });
+
+  channelMembers.map(member => {
+    channelMembersUid.push(member.data.user_id || '');
+  });
 
-  const channelMembers = channelMembersCollection.useWatcher({});
+  useChannelListener(channelMembersUid || []);
```
Comment (RomaricMourgues) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx:42]: Rename to useUsersListener
(and also rename the function file and folder)

**Comment by chamerling** [comment]
Can you put a GIF of the feature?

**Comment by RkAiman** [comment]
(1) Twake

**Comment by chamerling** [comment]
I do not like when the modal is changing its height when result changes. We should have a fixed height and some scroll. I think we did it already somewhere @RomaricMourgues @stephanevieira75 no?

**Comment by stephanevieira75** [comment]
> I do not like when the modal is changing its height when result changes. We should have a fixed height and some scroll. I think we did it already somewhere @RomaricMourgues @stephanevieira75 no?

Yes we use [react perfect scrollbar](https://www.npmjs.com/package/react-perfect-scrollbar)

```js
import PerfectScrollbar from 'react-perfect-scrollbar';

<PerfectScrollbar options={{ suppressScrollX: true }} component="div">
 ....
</PerfectScrollbar>
```

**Comment by RomaricMourgues** [comment]
@RkAiman : I agree with @chamerling and @stephanevieira75 
Can you have a static height ?
Everything else is veeeery good thanks a lot

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersEditor.tsx [hunk 4]
```diff
@@ -20,21 +27,29 @@ type Props = {
...
+        scope: 'group',
+        workspace_id: Workspaces.currentWorkspaceId,
+        group_id: Workspaces.currentGroupId,
+      },
+      (res: UserType[]) => {
+        res = res.filter(user => {
+          return membersIds.indexOf(user.id || '') < 0;
+        }, membersIds);
+        setMemberList(res);
+      },
```
Comment (chamerling) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersEditor.tsx:51]: Can be cleaner (personal taste):

```js
(res: UserType[]) => setMemberList(
  res.filter(user => membersIds.indexOf(user.id || "") < 0, membersIds)
)
```

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx [hunk 5]
```diff
@@ -1,50 +1,70 @@
...
 
+  const [search, setSearch] = useState('');
+  const [limit, setLimit] = useState(defaultLimit);
+  const [searchedUsers, setSearchedUsers] = useState<string[]>([]);
+
   const collectionPath: string = `/channels/v1/companies/${company_id}/workspaces/${workspace_id}/channels/${id}/members/`;
   const channelMembersCollection = Collections.get(collectionPath, ChannelMemberResource);
+  const channelMembers = channelMembersCollection.useWatcher({}, { limit: limit });
+
+  channelMembers.map(member => {
```
Comment (chamerling) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx:40]: This is not a `map` but a `forEach`. Keep map when you need to return a value from another one.
If `channelMembersUid` is only generated from channelMembers and is reactive, then you can use a map but like 

```js
const channelMembersUid = channelMembers.map(member => member.data.user_id)
```

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx [hunk 6]
```diff
@@ -1,50 +1,70 @@
...
         scope: 'workspace',
         workspace_id: workspace_id,
       },
-      (res: any) => {
-        callback(res);
+      (res: UserType[]) => {
+        const users: string[] = [];
+        res
+          .filter(user => channelMembersUid.includes(user.id || ''))
+          .map(user => users.push(user.id || ''));
```
Comment (chamerling) [inline twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx:58]: Same here, use map like 

```js
const users = res.filter().map()
```

Comment (RomaricMourgues) [reply twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx:58]:
```
 const users: string[] = res
          .filter(user => channelMembersUid.includes(user.id || ''))
          .map(user => user.id || '');
```

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersEditor.tsx (hunk 4 above)
Comment (RkAiman) [reply twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersEditor.tsx]: I agree. Thank you.

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx (hunk 5 above)
Comment (RkAiman) [reply twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx]: Done.

File: twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx (hunk 6 above)
Comment (RkAiman) [reply twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/ChannelMembersList.tsx]: Done.

**Comment by RkAiman** [comment]
@RomaricMourgues @chamerling Updated

(1) Twake

**Review by RomaricMourgues — APPROVED** [review]

**Pull request merged**